Python/Product/Analysis/Interpreter/Ast/AstAnalysisFunctionWalker.cs

```diff
@@ -111,25 +102,139 @@ FunctionDefinition targetFunction
+                    var typeId = BuiltinTypeIdFromName(typeName);
+                    if (typeId != BuiltinTypeId.Unknown) {
+                        _scope.SetInScope(name, new AstPythonConstant(new AstPythonBuiltinType(typeName, typeId)));
+                    }
+                }
+            }
             return base.Walk(node);
         }
 
         public override bool Walk(ReturnStatement node) {
-            foreach (var type in _scope.GetTypesFromValue(_scope.GetValueFromExpression(node.Expression))) {
-                _returnTypes.Add(type);
+            var typesAdded = false;
+            if (node.Expression is NameExpression nex && nex.Name == "self") {
+                // For self return the actual class without added private members
+                var klass = _scope.LookupNameInScopes("__class__", NameLookupContext.LookupOptions.Nonlocal);
+                if (klass != null) {
+                    _returnTypes.Add(klass as IPythonType);
+                    typesAdded = true;
+                }
+            }
+
+            if (!typesAdded) {
+                foreach (var type in _scope.GetTypesFromValue(_scope.GetValueFromExpression(node.Expression))) {
+                    _returnTypes.Add(type);
+                }
+            }
+            return true; // We want to evaluate all code so all private variables in __new__ get defined
+        }
+
+        private IMember GetClass() {
+            bool classmethod, staticmethod;
+            GetMethodType(_target, out classmethod, out staticmethod);
+            var klass = _scope.LookupNameInScopes("__class__", NameLookupContext.LookupOptions.Local);
+            if (!staticmethod && !classmethod) {
+                var cls = klass as IPythonType;
+                if (cls == null) {
+                    klass = null;
+                } else {
+                    klass = new AstPythonConstant(cls, ((cls as ILocatedMember)?.Locations).MaybeEnumerate().ToArray());
+                }
+            }
+            return klass;
+        }
+
+        private AstPythonType GetClassType(IMember klass) {
+            var cls = (klass as AstPythonConstant)?.Type as AstPythonType;
+            if (cls != null) {
+                var self = _scope.LookupNameInScopes("self", NameLookupContext.LookupOptions.Local);
+                if (self == null) {
+                    // Clone type since function analysis can add members that should not be
+                    // visible to the user such as private variables backing public properties.
+                    self = new AstPythonConstant(cls.Clone(), Array.Empty<LocationInfo>());
+                    _scope.SetInScope("self", self, mergeWithExisting: false);
+                }
+                return (self as AstPythonConstant)?.Type as AstPythonType;
             }
+            return null;
+        }
 
-            return false;
+        private static BuiltinTypeId BuiltinTypeIdFromName(string name) {
```

Comment:
Can we use `BuiltinTypeIdExtensions.GetTypeName` as the canonical source for these? Maybe even if we have to iterate through the values to get to them.

Reply:
Or just put a complete method next to the other one so that it's all there.

Reply:
I'll try just to look the built-in type up in scope.